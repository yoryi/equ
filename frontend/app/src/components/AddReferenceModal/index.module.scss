:local(.modalMessage) {
  font-family: Ilisarniq, sans-serif;
  font-size: 18px;
  line-height: 28px;
  color: var(--dark-grey);
  text-align: center;
}