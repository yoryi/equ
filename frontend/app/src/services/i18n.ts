import filesize from "filesize"
import i18n from "i18next"
import _ from "lodash"
import { initReactI18next } from "react-i18next"

import { en } from "../translations"

void i18n.use(initReactI18next).init({
  ns: [`common`, `screens`],
  defaultNS: `screens`,
  resources: {
    en,
  },
  lng: `en`,
  fallbackLng: `en`,
  interpolation: {
    format: (value, format, lng) => {
      if (!format) {
        return value
      }

      return format.split(`,`).reduce((value: string, format: string) => {
        switch (format.trim()) {
          case `localeString`:
            return value.toLocaleString()
          case `enIndefiniteArticle`:
          case `enIndefiniteArticleCapitalized`: {
            if (!lng || lng !== `en`) {
              return value
            }

            const letters = [`a`, `e`, `i`, `o`, `u`, `h`]
            const firstLetter = value.substring(0, 1)
            let article = ``
            if (
              letters.find((letter) => firstLetter.toLowerCase() === letter)
            ) {
              article =
                format === `enIndefiniteArticleCapitalized` ? `An` : `an`
            } else {
              article = format === `enIndefiniteArticleCapitalized` ? `A` : `a`
            }

            return `${article} ${value}`
          }
          case `title`: {
            return value
              .toLowerCase()
              .split(` `)
              .map((word: string) => _.upperFirst(word))
              .join(` `)
          }
          case `lowerCase`: {
            return value.toLowerCase()
          }
          case `fileSize`: {
            return filesize(parseInt(value), { round: 0 })
          }
          default:
            return value
        }
      }, value)
    },
  },
})

export default i18n
