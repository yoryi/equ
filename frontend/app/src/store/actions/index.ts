export * from "./admin"
export * from "./attachments"
export * from "./auth"
export * from "./follow"
export * from "./help"
export * from "./journeys"
export * from "./media"
export * from "./notifications"
export * from "./profile"
export * from "./searchAndFilters"
export * from "./settings"
export * from "./shared"
export * from "./student"
export * from "./surveys"
export * from "./university"
