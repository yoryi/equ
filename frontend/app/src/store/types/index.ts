export * from "./admin"
export * from "./auth"
export * from "./errors"
export * from "./ethnicity"
export * from "./follow"
export * from "./gender"
export * from "./hardship"
export * from "./help"
export * from "./highSchool"
export * from "./interest"
export * from "./notifications"
export * from "./profile"
export * from "./race"
export * from "./search"
export * from "./settings"
export * from "./shared"
export * from "./store"
export * from "./student"
export * from "./survey"
export * from "./university"
export * from "./utils"
