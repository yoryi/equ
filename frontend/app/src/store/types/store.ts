import { AdminReducer } from "../reducers/admin/reducer"
import { AuthReducer } from "../reducers/auth/reducer"
import { FollowReducer } from "../reducers/follow"
import { MediaReducer } from "../reducers/media/reducer"
import { NotificationsReducer } from "../reducers/notifications"
import { ProfileReducer } from "../reducers/profile/reducer"
import { SearchAndFiltersReducer } from "../reducers/searchAndFilters/reducer"
import { SettingsReducer } from "../reducers/settings/reducer"
import { SharedReducer } from "../reducers/shared/reducer"
import { StudentReducer } from "../reducers/student"
import { SurveysReducer } from "../reducers/surveys/reducer"
import { UniversityReducer } from "../reducers/university"

export interface ReduxState {
  auth: AuthReducer
  media: MediaReducer
  notifications: NotificationsReducer
  profile: ProfileReducer
  shared: SharedReducer
  settings: SettingsReducer
  student: StudentReducer
  surveys: SurveysReducer
  university: UniversityReducer
  follow: FollowReducer
  searchAndFilters: SearchAndFiltersReducer
  admin: AdminReducer
}
