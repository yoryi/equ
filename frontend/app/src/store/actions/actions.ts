/* eslint-disable @typescript-eslint/ban-types */
import { Action } from "redux"

import * as Types from "../types"

export interface PayloadAction<T> extends Action {
  payload: T
}

export interface PromiseAction<T> extends PayloadAction<T> {
  meta: {
    promise: {
      resolve: Function
      reject: Function
    }
  }
  then: Function
  catch: Function
  finally: Function
}

export interface ErrorAction extends Action {
  err: Error
}

export interface APIErrorAction extends Action {
  err: Types.APIError
}
