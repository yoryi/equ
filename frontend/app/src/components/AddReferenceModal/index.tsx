import { useFormik } from "formik"
import * as React from "react"
//Hooks
import { useEffect,useState } from "react"
//Services
import toast from "react-hot-toast"
import { useTranslation } from "react-i18next"
import { useDispatch } from "react-redux"
//Utils
import * as yup from "yup"

//Actions
import * as actions from "../../store/actions"
//Types
import { Activity,ReferenceType } from "../../store/types"
//Components
import Modal from "../Modal/Modal"
import TextField from "../TextField/TextField"
//Styles
import Styles from "./index.module.scss"

interface AddReferenceModalProps {
  visible: boolean
  onClose: () => void
}

type AddAcademicReferenceModalProps = {
  referenceType: ReferenceType.Academic
  activity?: undefined
} & AddReferenceModalProps

type AddActivityReferenceModalProps = {
  referenceType: ReferenceType.Activity
  activity: Activity
} & AddReferenceModalProps

interface FormValues {
  email: string
}

const AddReferenceModal: React.FC<
  AddAcademicReferenceModalProps | AddActivityReferenceModalProps
> = ({ visible, onClose, referenceType, activity }) => {
  const dispatch = useDispatch()

  const [isSuccessModalVisible, setSuccessModalVisible] = useState(false)

  const { t } = useTranslation()

  const validationSchema = yup.object({
    email: yup
      .string()
      .email(t(`errors:invalidEmailAddress`))
      .required(t(`errors:cannotBeEmpty`, { name: t(`common:email`) })),
  })

  const {
    values,
    touched,
    errors,
    handleChange,
    handleSubmit,
    isSubmitting,
    resetForm,
  } = useFormik<FormValues>({
    initialValues: {
      email: ``,
    },
    validationSchema,
    onSubmit: async ({ email }, { setErrors }) => {
      try {
        await dispatch(
          actions.sendReferenceLink(
            referenceType === ReferenceType.Academic
              ? {
                  type: ReferenceType.Academic,
                  email,
                }
              : { type: ReferenceType.Activity, email, activity: activity! },
          ),
        )

        onClose()
        setSuccessModalVisible(true)
      } catch (err) {
        if (err.message) {
          toast.error(err.message)
        }

        setErrors(err.errors ?? {})
      }
    },
  })

  useEffect(() => {
    resetForm()
  }, [resetForm, visible])

  const handleSuccessModalClose = () => setSuccessModalVisible(false)

  return (
    <>
      <Modal
        title={t(`profile.student.transcriptScreen.requestAcademicReference`)}
        isOpen={visible}
        toggle={onClose}
        footer={t(`profile.student.transcriptScreen.requestReference`)}
        footerAction={handleSubmit}
        footerButtonAriaLabel={`submit`}
        disabledFooterButton={isSubmitting}
        closeIcon
      >
        <TextField
          id={`email`}
          name={`email`}
          title={t(`profile.student.transcriptScreen.enterEmailAddressBelow`)}
          value={values.email}
          onChange={handleChange}
          error={touched.email && errors.email}
          size={`lg`}
        />
      </Modal>

      <Modal
        isOpen={isSuccessModalVisible}
        toggle={handleSuccessModalClose}
        title={t(`common:success`)}
        size={`sm`}
        footer={t(`common:continue`)}
        footerAction={handleSuccessModalClose}
        footerStyle={{ marginBottom: 32 }}
        closeIcon
      >
        <p className={Styles.modalMessage}>
          {t(`common:referenceRequestedSuccessfully`)}
        </p>
      </Modal>
    </>
  )
}

export default AddReferenceModal
