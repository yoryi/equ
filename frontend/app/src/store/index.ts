import * as actions from "./actions"
import { initReduxStore } from "./config"
import { store } from "./createStore"

export { actions,initReduxStore, store }
