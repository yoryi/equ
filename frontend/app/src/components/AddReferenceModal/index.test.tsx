// @ts-ignore
import { promiseMiddleware } from "@adobe/redux-saga-promise"
//Testing
import { act, fireEvent,render } from "@testing-library/react"
import faker from "faker"
import _ from "lodash"
import * as React from "react"
//Services
import toast from "react-hot-toast"
//Translations
import { I18nextProvider } from "react-i18next"
//Redux store
import { Provider } from "react-redux"
import configureStore from "redux-mock-store"

import i18n from "../../services/i18n"
import * as actions from "../../store/actions"
//Types
import { Activity, ActivityType,ReferenceType } from "../../store/types"
//Components
import AddReferenceModal from "./"

describe(`AddReferenceModal component`, () => {
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  const toastSpy = jest
    .spyOn(toast, `error`)
    .mockImplementation((() => {}) as any)

  const mockStore = configureStore([promiseMiddleware])
  const store = mockStore()

  const wrapper: React.FC = ({ children }) => (
    <I18nextProvider i18n={i18n}>
      <Provider store={store}>{children}</Provider>
    </I18nextProvider>
  )

  afterEach(() => {
    store.clearActions()
    toastSpy.mockReset()
  })

  it(`should render and match the snapshot`, () => {
    const { container } = render(
      <AddReferenceModal
        onClose={jest.fn()}
        referenceType={ReferenceType.Academic}
        visible
      />,
      {
        wrapper,
      },
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  describe(`should submit the form`, () => {
    let email: string
    beforeEach(() => {
      email = faker.internet.email()
    })

    describe(`academic reference`, () => {
      it(`success`, async () => {
        const action = actions.sendReferenceLink({
          type: ReferenceType.Academic,
          email,
        })
        const onClose = jest.fn()

        const { getByLabelText } = render(
          <AddReferenceModal
            onClose={onClose}
            referenceType={ReferenceType.Academic}
            visible
          />,
          { wrapper },
        )

        act(() => {
          fireEvent.change(
            getByLabelText(
              i18n.t(
                `screens:profile.student.transcriptScreen.enterEmailAddressBelow`,
              ),
            ),
            {
              target: { value: email },
            },
          )
        })

        await act(async () => {
          fireEvent.click(getByLabelText(`submit`))
        })

        await act(async () => {
          store
            .getActions()
            .find(({ type }) => type === action.type)
            ?.meta.promise.resolve()
        })

        expect(
          store.getActions().find(({ type }) => type === action.type).payload,
        ).toEqual(action.payload)
        expect(onClose).toHaveBeenCalled()
      })

      describe(`failure`, () => {
        it(`validation error`, async () => {
          const action = actions.sendReferenceLink({
            type: ReferenceType.Academic,
            email,
          })
          const onClose = jest.fn()
          const errors = {
            email: faker.lorem.sentence(),
          }

          const { getByLabelText, queryByText } = render(
            <AddReferenceModal
              onClose={onClose}
              referenceType={ReferenceType.Academic}
              visible
            />,
            { wrapper },
          )

          act(() => {
            fireEvent.change(
              getByLabelText(
                i18n.t(
                  `screens:profile.student.transcriptScreen.enterEmailAddressBelow`,
                ),
              ),
              {
                target: { value: email },
              },
            )
          })

          await act(async () => {
            fireEvent.click(getByLabelText(`submit`))
          })

          await act(async () => {
            store
              .getActions()
              .find(({ type }) => type === action.type)
              ?.meta.promise.reject({ errors })
          })

          expect(
            store.getActions().find(({ type }) => type === action.type).payload,
          ).toEqual(action.payload)
          expect(queryByText(errors.email)).not.toBeNull()
          expect(onClose).not.toHaveBeenCalled()
        })

        it(`message error`, async () => {
          const action = actions.sendReferenceLink({
            type: ReferenceType.Academic,
            email,
          })
          const onClose = jest.fn()
          const message = faker.lorem.sentence()

          const { getByLabelText } = render(
            <AddReferenceModal
              onClose={onClose}
              referenceType={ReferenceType.Academic}
              visible
            />,
            { wrapper },
          )

          act(() => {
            fireEvent.change(
              getByLabelText(
                i18n.t(
                  `screens:profile.student.transcriptScreen.enterEmailAddressBelow`,
                ),
              ),
              {
                target: { value: email },
              },
            )
          })

          await act(async () => {
            fireEvent.click(getByLabelText(`submit`))
          })

          await act(async () => {
            store
              .getActions()
              .find(({ type }) => type === action.type)
              ?.meta.promise.reject({ message })
          })

          expect(
            store.getActions().find(({ type }) => type === action.type).payload,
          ).toEqual(action.payload)
          expect(toastSpy).toHaveBeenCalledWith(message)
          expect(onClose).not.toHaveBeenCalled()
        })
      })
    })

    describe(`activity reference`, () => {
      let activity: Activity
      beforeEach(() => {
        activity = {
          id: faker.random.number(10000),
          title: faker.lorem.words(2),
          startDate: null,
          endDate: null,
          company: faker.company.companyName(),
          description: faker.lorem.sentences(2),
          type: _.sample([
            ActivityType.EMPLOYMENT,
            ActivityType.INTERNSHIP,
          ]) as ActivityType,
          referenceLink: null,
        }
      })

      it(`success`, async () => {
        const action = actions.sendReferenceLink({
          type: ReferenceType.Activity,
          email,
          activity,
        })
        const onClose = jest.fn()

        const { getByLabelText, getByText } = render(
          <AddReferenceModal
            onClose={onClose}
            referenceType={ReferenceType.Activity}
            activity={activity}
            visible
          />,
          { wrapper },
        )

        act(() => {
          fireEvent.change(
            getByLabelText(
              i18n.t(
                `screens:profile.student.transcriptScreen.enterEmailAddressBelow`,
              ),
            ),
            {
              target: { value: email },
            },
          )
        })

        await act(async () => {
          fireEvent.click(getByLabelText(`submit`))
        })

        await act(async () => {
          store
            .getActions()
            .find(({ type }) => type === action.type)
            ?.meta.promise.resolve()
        })

        expect(
          store.getActions().find(({ type }) => type === action.type).payload,
        ).toEqual(action.payload)
        expect(onClose).toHaveBeenCalled()

        await act(async () => {
          fireEvent.click(getByText(i18n.t(`common:continue`)))
        })

        // TODO Fix failing expect
        // expect(queryAllByRole('dialog').length).toEqual(1);
      })

      describe(`failure`, () => {
        it(`validation error`, async () => {
          const action = actions.sendReferenceLink({
            type: ReferenceType.Activity,
            email,
            activity,
          })
          const onClose = jest.fn()
          const errors = {
            email: faker.lorem.sentence(),
          }

          const { getByLabelText, queryByText } = render(
            <AddReferenceModal
              onClose={onClose}
              referenceType={ReferenceType.Activity}
              activity={activity}
              visible
            />,
            { wrapper },
          )

          act(() => {
            fireEvent.change(
              getByLabelText(
                i18n.t(
                  `screens:profile.student.transcriptScreen.enterEmailAddressBelow`,
                ),
              ),
              {
                target: { value: email },
              },
            )
          })

          await act(async () => {
            fireEvent.click(getByLabelText(`submit`))
          })

          await act(async () => {
            store
              .getActions()
              .find(({ type }) => type === action.type)
              ?.meta.promise.reject({ errors })
          })

          expect(
            store.getActions().find(({ type }) => type === action.type).payload,
          ).toEqual(action.payload)
          expect(queryByText(errors.email)).not.toBeNull()
          expect(onClose).not.toHaveBeenCalled()
        })

        it(`message error`, async () => {
          const action = actions.sendReferenceLink({
            type: ReferenceType.Activity,
            email,
            activity,
          })
          const onClose = jest.fn()
          const message = faker.lorem.sentence()

          const { getByLabelText } = render(
            <AddReferenceModal
              onClose={onClose}
              referenceType={ReferenceType.Activity}
              activity={activity}
              visible
            />,
            { wrapper },
          )

          act(() => {
            fireEvent.change(
              getByLabelText(
                i18n.t(
                  `screens:profile.student.transcriptScreen.enterEmailAddressBelow`,
                ),
              ),
              {
                target: { value: email },
              },
            )
          })

          await act(async () => {
            fireEvent.click(getByLabelText(`submit`))
          })

          await act(async () => {
            store
              .getActions()
              .find(({ type }) => type === action.type)
              ?.meta.promise.reject({ message })
          })

          expect(
            store.getActions().find(({ type }) => type === action.type).payload,
          ).toEqual(action.payload)
          expect(toastSpy).toHaveBeenCalledWith(message)
          expect(onClose).not.toHaveBeenCalled()
        })
      })
    })
  })
})
