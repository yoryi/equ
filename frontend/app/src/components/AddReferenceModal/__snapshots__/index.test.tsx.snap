// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddReferenceModal component should render and match the snapshot 1`] = `
<div
  class="modal-backdrop fade"
/>
`;
